pysrc/pack.py

```diff
@@ -0,0 +1,594 @@
+                    node2_x = row["node2_x"]
+                    node1_y = row["node1_y"]
+                    node2_y = row["node2_y"]
+                    if node1_x != node2_x:
+                        raise AssertionError("x's must be the same")
+                    if abs(node1_y - node2_y) != 1:
+                        raise AssertionError("batteries can only take up one y")
+                    batt_y = min(node1_y, node2_y) + 0.5
+                    temperature = self.get_new_cell_temperature()
+                    self.batteries[desc].update(
+                        {"x": node1_x, "y": batt_y, "temperature": temperature}
+                    )
+                self.batteries[desc].update({"offset": self.offset})
+                self.offset += self.cell_size
+
+        if self._thermal:
+            self.build_thermal_equations()
+
+        self.num_cells = len(self.batteries)
+
+        if len(curr_sources) != 1:
+            raise NotImplementedError("can't do this yet")
+        # copy the basis which we can use to place the loop currents
+        basis_to_place = deepcopy(mcb)
+        self.place_currents(loop_currents, basis_to_place)
+        pack_eqs_vec = self.build_pack_equations(loop_currents, curr_sources)
+        self.len_pack_eqs = len(pack_eqs_vec)
+        pack_eqs = pybamm.numpy_concatenation(*pack_eqs_vec)
+
+        cells = [d["cell"] for d in self.batteries.values()]
+        cell_eqs = pybamm.numpy_concatenation(*cells)
+
+
+
+
+        self.pack = pybamm.numpy_concatenation(pack_eqs, cell_eqs)
+        if self._implicit:
+            len_sv = len(cells)*self.cell_size + len(pack_eqs_vec)
+            sv = pybamm.StateVector(slice(0,len_sv))
+            dsv = pybamm.StateVectorDot(slice(0,len_sv))
+            p = pybamm2julia.PsuedoInputParameter("lolol")
+            t = pybamm.Time()
+            self.pack = pybamm2julia.PybammJuliaFunction(
+                [dsv, sv, p, t], self.pack, "pack", True
+            )
+        self.ics = self.initialize_pack(num_loops, len(curr_sources))
+
+    def initialize_pack(self, num_loops, num_curr_sources):
+        curr_ics = pybamm.Vector([1.0 for curr_source in range(num_loops)])
+        curr_source_v_ics = pybamm.Vector(
+            [1.0 for curr_source in range(num_curr_sources)]
+        )
+        cell_ics = pybamm.numpy_concatenation(
+            *[
+                self.built_model.concatenated_initial_conditions
+                for n in range(len(self.batteries))
+            ]
+        )
+        ics = pybamm.numpy_concatenation(*[curr_ics, curr_source_v_ics, cell_ics])
+        return ics
+
+    def build_pack_equations(self, loop_currents, curr_sources):
+        # start by looping through the loop currents. Sum Voltages
+        pack_equations = []
+        cells = []
+        for i, loop_current in enumerate(loop_currents):
+            # loop through the edges
+            eq = []
+            for edge in self.circuit_graph.edges:
```

Comment:
would be more readable to loop over items (`for name, edge in self.circuit_graph.edges.items()`) instead of calling `self.circuit_graph.edges[edge]` each time